Non Terminating Worksheet Evaluation for SO Condition

**Describe the bug**
Code that returns an SO in a normal App causes a non-terminating `Evaluated Worksheet` message and does not return an SO.

**To Reproduce**
Below code should reproduce issue within a `worksheet.sc` file.

```
def bigData = (1 to 1000000).toStream
def foldLeft[A, B](l: List[A], z: B)(f: (B, A) => B): B = l match {
  case Nil    => z
  case h :: t => foldLeft(t, f(z, h))(f)
}
def foldRightViaFoldLeft[A, B](l: List[A], z: B)(s: (A, B) => B): B =
  foldLeft(l, (x: B) => x)((e, f) => (x: B) => e(s(f, x)))(z)
foldRightViaFoldLeft(bigData.toList, 0)(_ + _) // doesn't terminate
```

**Installation:**
- Operating system: Linux
- Editor: Visual Studio Code
- Metals version: 
```
metals.serverVersion": "0.8.0+294-4dcdec07-SNAPSHOT
```

Error thrown in this example is:
```
Caused by: java.lang.StackOverflowError
	at scala.collection.immutable.Nil$.equals(List.scala:435)
	at repl.Session$App.foldLeft(test.worksheet.sc:21)
	at repl.Session$App.foldLeft(test.worksheet.sc:22)
	at repl.Session$App.foldLeft(test.worksheet.sc:22)
	at repl.Session$App.foldLeft(test.worksheet.sc:22)
...
```

It is not caught here:
```
          try result.complete(Some(evaluateWorksheet(path, token)))
          catch {
            case e @ (NonFatal(_) | InterruptException()) =>
              result.completeExceptionally(e)
          }
```

Adding catching `e: java.lang.StackOverflowError => ...` solves a problem of infinite evaluation.
It might be changed to something more generic if needed of course.

But I can see here 2 problems:
1) With this change only there is no visual feedback to the user (in VS Code) that evaluation failed.
2) When evaluation fails it should clear previous evaluation results because it might be super misleading. E.g. when a code in worksheet is changed and no longer compiles previous evaluations are erased. I think it would make sense for similar thing in case of failed evaluation.

--- a/build.sbt
+++ b/build.sbt
@@ -155,7 +155,7 @@ lazy val V = new {
   val sbtBloop = bloop
   val gradleBloop = bloop
   val mavenBloop = bloop
-  val mdoc = "2.1.1"
+  val mdoc = "2.1.3"
   val scalafmt = "2.4.1"
   val munit = "0.4.3"
   // List of supported Scala versions in SemanticDB. Needs to be manually updated

--- a/metals/src/main/scala/scala/meta/internal/metals/Embedded.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/Embedded.scala
@@ -99,7 +99,7 @@ object Embedded {
         Repository.central(),
         Repository.ivy2Local(),
         MavenRepository.of(
-          "https://oss.sonatype.org/content/repositories/releases/"
+          "https://oss.sonatype.org/content/repositories/public/"
         ),
         MavenRepository.of(
           "https://oss.sonatype.org/content/repositories/snapshots/"

--- a/metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala
+++ b/metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala
@@ -23,7 +23,6 @@ import java.util.concurrent.TimeUnit
 import scala.meta.internal.metals.MutableCancelable
 import scala.meta.internal.metals.StatusBar
 import scala.meta.internal.pc.InterruptException
-import scala.util.control.NonFatal
 import java.util.concurrent.Executors
 import scala.meta.internal.metals.Diagnostics
 import scala.meta.internal.metals.Timer
@@ -120,10 +119,11 @@ class WorksheetProvider(
       }
     }
     val onError: PartialFunction[Throwable, Option[EvaluatedWorksheet]] = {
-      case NonFatal(e) =>
-        scribe.error(s"worksheet: $path", e)
-        None
       case InterruptException() =>
+        None
+      case e: Throwable =>
+        // NOTE(olafur): we catch all exceptions because of https://github.com/scalameta/metals/issues/1456
+        scribe.error(s"worksheet: $path", e)
         None
     }
     def runEvaluation(): Unit = {
@@ -145,11 +145,10 @@ class WorksheetProvider(
       // infinite loops.
       val thread = new Thread(s"Evaluating Worksheet ${path.filename}") {
         override def run(): Unit = {
-          try result.complete(Some(evaluateWorksheet(path, token)))
-          catch {
-            case e @ (NonFatal(_) | InterruptException()) =>
-              result.completeExceptionally(e)
-          }
+          result.complete(
+            try Some(evaluateWorksheet(path, token))
+            catch onError
+          )
         }
       }
       interruptThreadOnCancel(path, result, thread)

--- a/tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala
+++ b/tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala
@@ -383,4 +383,48 @@ abstract class BaseWorksheetLspSuite(scalaVersion: String)
     } yield ()
   }
 
+  test("fatal-exception") {
+    for {
+      _ <- server.initialize(
+        s"""
+           |/metals.json
+           |{"a": {"scalaVersion": "$scalaVersion"}}
+           |/a/src/main/scala/StackOverflowError.worksheet.sc
+           |throw new StackOverflowError()
+           |/a/src/main/scala/NoSuchMethodError.worksheet.sc
+           |throw new NoSuchMethodError()
+           |/a/src/main/scala/IncompatibleClassChangeError.worksheet.sc
+           |throw new IncompatibleClassChangeError()
+           |""".stripMargin
+      )
+      _ <- server.didOpen("a/src/main/scala/StackOverflowError.worksheet.sc")
+      _ <- server.didOpen(
+        "a/src/main/scala/IncompatibleClassChangeError.worksheet.sc"
+      )
+      _ <- server.didOpen("a/src/main/scala/NoSuchMethodError.worksheet.sc")
+      _ = assertNoDiff(
+        client.workspaceDiagnostics,
+        """|a/src/main/scala/IncompatibleClassChangeError.worksheet.sc:1:1: error: java.lang.IncompatibleClassChangeError
+           |	at repl.Session$App.<init>(IncompatibleClassChangeError.worksheet.sc:8)
+           |	at repl.Session$.app(IncompatibleClassChangeError.worksheet.sc:3)
+           |
+           |throw new IncompatibleClassChangeError()
+           |^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+           |a/src/main/scala/NoSuchMethodError.worksheet.sc:1:1: error: java.lang.NoSuchMethodError
+           |	at repl.Session$App.<init>(NoSuchMethodError.worksheet.sc:8)
+           |	at repl.Session$.app(NoSuchMethodError.worksheet.sc:3)
+           |
+           |throw new NoSuchMethodError()
+           |^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+           |a/src/main/scala/StackOverflowError.worksheet.sc:1:1: error: java.lang.StackOverflowError
+           |	at repl.Session$App.<init>(StackOverflowError.worksheet.sc:8)
+           |	at repl.Session$.app(StackOverflowError.worksheet.sc:3)
+           |
+           |throw new StackOverflowError()
+           |^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+           |""".stripMargin
+      )
+    } yield ()
+  }
+
 }

